shell/config/uiplugins.js

```diff
@@ -57,123 +50,132 @@ export const UI_PLUGIN_CHART_ANNOTATIONS = {
+ * Get value of the annotation from the latest version for a chart
  */
 export function uiPluginAnnotation(chart, name) {
-  if (chart?.versions?.length > 0) {
-    return chart.versions[0].annotations?.[name];
-  }
-
-  return undefined;
+  return chart?.versions?.[0]?.annotations?.[name];
 }
 
 /**
- * Parse the rancher version string
+ * Parse the Rancher version string
  */
 function parseRancherVersion(v) {
-  let parsedRancherVersion = semver.coerce(v)?.version;
-  const splitArr = parsedRancherVersion.split('.');
+  let parsedVersion = semver.coerce(v)?.version;
+  const splitArr = parsedVersion?.split('.');
 
   // this is a scenario where we are on a "head" version of some sort... we can't infer the patch version from it
   // so we apply a big patch version number to make sure we follow through with the minor
   if (v.includes('-') && splitArr?.length === 3) {
-    parsedRancherVersion = `${ splitArr[0] }.${ splitArr[1] }.999`;
+    parsedVersion = `${ splitArr[0] }.${ splitArr[1] }.999`;
   }
 
-  return parsedRancherVersion;
+  return parsedVersion;
 }
 
-// i18n-uses plugins.error.generic, plugins.error.api, plugins.error.host
+/**
+ * Check if a version is incompatible with the current environment
+ */
+function checkIncompatibility(currentVersion, requiredVersion, incompatibilityData, returnObj, versionObj) {
+  if (requiredVersion && !semver.satisfies(currentVersion, requiredVersion)) {
+    if (!returnObj) {
+      return false;
+    }
+    versionObj.isVersionCompatible = false;
+    versionObj.versionIncompatibilityData = { ...incompatibilityData, required: requiredVersion };
+
+    return versionObj;
+  }
+
+  return true;
+}
 
 /**
  * Whether an extension should be loaded based on the metadata returned by the backend in the UIPlugins resource instance
+ * The output will be used to PREVENT loading of an extension that is already installed but isn't compatible with the system
  * @returns String || Boolean
  */
-export function shouldNotLoadPlugin(UIPluginResource, rancherVersion, loadedPlugins) {
+export function shouldNotLoadPlugin(UIPluginResource, { rancherVersion, kubeVersion }, loadedPlugins) {
   if (!UIPluginResource.name || !UIPluginResource.version || !UIPluginResource.endpoint) {
     return 'plugins.error.generic';
   }
 
   // Extension chart specified a required extension API version
   // we are propagating the annotations in pkg/package.json for any extension
   // inside the "spec.plugin.metadata" property of UIPlugin resource
-  const requiredUiExtensionsVersion = UIPluginResource.spec?.plugin?.metadata?.[UI_PLUGIN_CHART_ANNOTATIONS.EXTENSIONS_VERSION];
+  // the minimum version is enforced to prevent loading of extensions
+  // that don't have the annotation 'catalog.cattle.io/ui-extensions-version' present
+  const requiredUiExtensionsVersion = UIPluginResource.spec?.plugin?.metadata?.[UI_PLUGIN_CHART_ANNOTATIONS.EXTENSIONS_VERSION] || MINIMUM_MANDATORY_EXTENSIONS_API_VERSION;
```

Comment:
Are we sure this is right? We're falling back to the minimum required version being the current shell package rather than a hardcoded 3.0.0

Reply:
We can go both ways here:

Either we hardcode it to `>= 3.0.0`, but have to be mindful to change this in the future whenever we change the `UI_EXTENSIONS_API_VERSION` or we keep it as I've done where we don't need to worry about it `>= ${ UI_EXTENSIONS_API_VERSION }`.

The goal here is to prevent loading of an extension that doesn't have the annotation `catalog.cattle.io/ui-extensions-version`, right @nwmac ?

If you think we should hardcode it, it's a 30sec fix but there's a drawback that you should be aware.

Reply:
The question is should it be 3.0.0 forever, or should it always be the shell that rancher was shipped with. I thought it was the former, but i can see how the later would force people to set an annotation.

Reply:
Chatted with Neil (see slack).

If the annotation is missing we should assume it's `< 3.0.0` (i.e not compatible with the vue3 bump)

Reply:
Ok, got to the bottom comment.

So now if the annotation is missing, in a shell 3.0.0+ world, we won't load the extension and instead show a specific warning. Suggestion on that wording incoming, but overall good solution